hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java

```diff
@@ -1130,17 +1359,23 @@ private void blockRootDelete(String key) throws InvalidRequestException {
    * Perform a bulk object delete operation.
    * Increments the {@code OBJECT_DELETE_REQUESTS} and write
    * operation statistics.
+   * Retry policy: retry untranslated; delete considered idempotent.
    * @param deleteRequest keys to delete on the s3-backend
    * @throws MultiObjectDeleteException one or more of the keys could not
    * be deleted.
    * @throws AmazonClientException amazon-layer failure.
    */
+  @Retries.RetryRaw
   private void deleteObjects(DeleteObjectsRequest deleteRequest)
-      throws MultiObjectDeleteException, AmazonClientException {
+      throws MultiObjectDeleteException, AmazonClientException, IOException {
     incrementWriteOperations();
-    incrementStatistic(OBJECT_DELETE_REQUESTS, 1);
     try {
-      s3.deleteObjects(deleteRequest);
+      invoker.retryUntranslated("delete",
```

Comment:
OK. How about: no retry/untranslated in innerDelete, -> retryTranslated in API delete, with anything the committer doing wrapping it in its own retries and/or ignoring of failures (which of course, is  a valid option in some cases)

Reply:
not quite following your comment here.  One question on my mind is the behavior of recursive delete.  Should it fail if the set of objects changes underneath it during execution?  For example, I'm executing a recursive delete, and an object disappears due to concurrent modification or eventual consistency.  Should the delete fail, or succeed because the contract of delete(path) was satisfied (stuff was deleted).  Based on my reading of your RetryPolicy, you don't retry on object not found (good), but we still fail.  We may want to actually throw to top-level delete() and restart the recursive delete.  A bit orthogonal but I want to understand the big picture to feel confident this code is right, and where it is headed in the future.

Reply:
Fundamentally, the core contract of a filesystem directory delete: it changes from the state "is" to the state "isnt", in an O(1) atomic op, is not satisified, we just pretend it does..

There's already ambiguity in our pretence. If an object disappears during a delete, that is not detected. If a new object is added if it is in a large dir and a later list() call finds it, it is deleted. Otherwise, it remains. Which means that we don't even detect updates to a path, not unless after a delete() you do another list to see if there has been a change (or do some leasing in dynamo, but let's not go there).

AFAIK, delete() succeeds if any child is taken away; the outcome on an add is undefined. rename is a bit more ambiguous w.r.t either.